### PR TITLE
Down special air special button variability

### LINKED ISSUE #10: Down Special (Air): Change behavior depending on special button press.
Current state: Once down special activates, Ganondorf will dive down until he hits a solid object or dies. 

Desired State: 
- As long as the special button is held down during the down special in the air, have Ganondorf act in the current state. 
- If the special button is not pressed, Ganondorf will conduct his current state for only 10 frames. 

On the 11th frame, provided Ganondorf is still in the air: 
1. A weaker hitbox will replace the hand hitbox. 
2. Will slow down for 20 frames until y speed approaches -0.0. 
3. If Ganondorf touches the ground, transition to landing animation with NO landing hitboxes. Else, transition to fall animation.

### PR DESCRIPTION
**Pull request by ZeDespo**
Buckle up, this is a big one. 

If Ganondorf's down special air has the special button held, we get a nice paralyzing hitbox into an explosion if we hit the ground. 

If the special button is let go, Ganondorf will...
1. eventually bring Ganondorf to a halt if still in the air after 25 frames. 
2. nerf the damage and knockback if in this slowdown state. 
3. will not transition to an explosion, but rather a landing, if ganondorf hits the ground.

### CODE REVIEW

File: src/ganon/acmd/specials/special_air_side_fall.rs [hunk 1]
```diff
@@ -1,99 +1,137 @@
...
+            collision = Hash40::new("collision_attr_purple");
+        }
+        if macros::is_excute(agent) {
+            if i == 2 {
+                WorkModule::set_int(
+                    agent.module_accessor,
+                    *FIGHTER_GANON_EXPLOSION_FALL_SETTING_FALL,
+                    *FIGHTER_GANON_STATUS_WORK_ID_INT_EXPLOSION_FALL_SETTING,
+                );
+            } else if i == 19 && !use_weak_hitbox(agent) {
```
Comment (FatherOfEgg) [inline src/ganon/acmd/specials/special_air_side_fall.rs:50]: The variable `i` will never reach 19.

File: src/ganon/status/down_special.rs [hunk 2]
```diff
@@ -1,17 +1,144 @@
...
+    let fvar6 = WorkModule::get_param_float(
+        boma,
+        0xfea97fe73,
+        hash40("special_s_fall_check_dead_offset_y"),
+    );
+    WorkModule::set_float(
+        boma,
+        fvar6,
+        *FIGHTER_INSTANCE_WORK_ID_FLOAT_CHECK_DEAD_OFFSET_Y,
+    );
```
Comment (FatherOfEgg) [inline src/ganon/status/down_special.rs]: This is unnecessary because this is to make Ganon die first when grabbing someone, and they're falling towards the bottom blast zone. Ganon doesn't have anyone in his grasp, so we don't need it.

Also, that hash, `0xfea97fe73`, is `hash40("param_special_s")`

File: src/ganon/status/down_special.rs [hunk 3]
```diff
@@ -1,17 +1,144 @@
...
+                    KineticModule::clear_speed_all(boma);
+                    KineticModule::change_kinetic(boma, *FIGHTER_KINETIC_TYPE_MOTION_FALL);
+                    fighter.change_status(FIGHTER_STATUS_KIND_FALL.into(), false.into());
+                } else {
+                    fighter.change_status(FIGHTER_STATUS_KIND_LANDING.into(), false.into());
+                }
+                return 0.into();
+            }
+            if !AttackModule::is_infliction(boma, *COLLISION_KIND_MASK_HIT) {
+                WorkModule::set_int(boma, countdown - 1, GANON_DOWN_SPECIAL_AIR_COUNTDOWN_INT);
```
Comment (FatherOfEgg) [inline src/ganon/status/down_special.rs]: This can be:
`WorkModule::dec_int(boma, GANON_DOWN_SPECIAL_AIR_COUNTDOWN_INT);`

File: src/ganon/status/down_special.rs [hunk 4]
```diff
@@ -1,17 +1,144 @@
...
+                    set_speed,
+                    fighter,
+                    FIGHTER_KINETIC_ENERGY_ID_GRAVITY,
+                    KineticModule::get_sum_speed_y(boma, *KINETIC_ENERGY_RESERVE_ATTRIBUTE_MAIN)
+                        / 1.9
+                );
+            } else if countdown == 0 {
+                if boma.is_situation(*SITUATION_KIND_AIR) {
+                    KineticModule::clear_speed_all(boma);
+                    KineticModule::change_kinetic(boma, *FIGHTER_KINETIC_TYPE_MOTION_FALL);
```
Comment (FatherOfEgg) [inline src/ganon/status/down_special.rs]: `KineticModule::change_kinetic(boma, *FIGHTER_KINETIC_TYPE_MOTION_FALL);` doesn't need to be here, since we're changing into the fall status.

File: src/ganon/acmd/specials/special_air_side_fall.rs [hunk 5]
```diff
@@ -1,99 +1,137 @@
...
+            damage = 14.0;
+            kbg = 113;
+            bkb = 60;
+            collision = Hash40::new("collision_attr_paralyze");
+        } else {
+            damage = 10.0;
+            kbg = 86;
+            bkb = 50;
+            collision = Hash40::new("collision_attr_purple");
+        }
```
Comment (FatherOfEgg) [inline src/ganon/acmd/specials/special_air_side_fall.rs:42]: It appears that this is always doing the strong hitbox. I never was able to get the non-paralyze hitbox.

File: src/ganon/acmd/specials/special_air_side_fall.rs [hunk 6]
```diff
@@ -1,99 +1,137 @@
...
+            WorkModule::get_int(agent.module_accessor, GANON_DOWN_SPECIAL_AIR_COUNTDOWN_INT);
+        return countdown > 0 && countdown <= 10;
+    }
+    false
+}
+
+/// specialairsfall is on a 19 frame cycle. At any time in those 19 frames, the player
+/// can release the special button to transition into slowdown state of down special air,
+/// and should be reflected in the hitboxes.
 unsafe extern "C" fn ganon_specialairsfall(agent: &mut L2CAgentBase) {
```
Comment (FatherOfEgg) [inline src/ganon/acmd/specials/special_air_side_fall.rs:22]: Is it fine that once the "19" (18) frame window is over, it maintains the strength of the hitboxes even if you let go of the special button? For example, if you hold the special button, and you let go of the button once those "19" frames are up, it will still be the strong hitbox and he will then transition into the slowdown state.

**Review by FatherOfEgg — CHANGES_REQUESTED** [review]

File: src/ganon/acmd/specials/special_air_side_fall.rs (hunk 6 above)
Comment (ZeDespo) [reply src/ganon/acmd/specials/special_air_side_fall.rs:22]: Addressed! Once the "weak" hitbox happens, we will not trigger the damage multiplier

File: src/ganon/acmd/specials/special_air_side_fall.rs (hunk 5 above)
Comment (ZeDespo) [reply src/ganon/acmd/specials/special_air_side_fall.rs:42]: Fixed this with something absolutely evil. Check your discord!

File: src/ganon/acmd/specials/special_air_side_fall.rs
Comment (FatherOfEgg) [inline src/ganon/acmd/specials/special_air_side_fall.rs:1]: I feel like this acmd could be reworked a lot:

1. It is really difficult to hit with the weak hitbox (the non-paraylyze/`collision_attr_purple` one)
 a.  You have to tap the special button and no longer. Literally only 1 frame just to get into the move and don't hold it anymore
 b. If you let go of the special button on frame 2, it is not possible to get the weak hitbox (non-paralyze/`collision_attr_purple`), because `GANON_DOWN_SPECIAL_AIR_COUNTDOWN_INT` will not be within the range 0 to 10
 c. You have to be high enough to let `GANON_DOWN_SPECIAL_AIR_COUNTDOWN_INT` decrement until it's within 0 and 10
 d. Essentially, you can only hit the weak hitbox (non-paralyze/`collision_attr_purple`) near the tail end of it **BEFORE** you transition to the FALL status
 e. More of this is explained in the later points
2. If you can hold the special button and fall infinitely
 a. You will go beyond the 19 frame window
 b. It will always be the strong hitbox (paralyze)
 c. Letting go of the special button frame 2 and onward will always make it be the strong hitbox
3. I feel like some of the things in the for loop aren't necessary and can be reworked to have only the necessary things be in there and make it more concise.
 a. When `i == 2`, does that really need to be there?
 b. The same thing sort of applies to when `i == 19 && ...`

Is this the intended mechanic? Pressing/holding the special button for no longer than 1 frame, **AND** have to be high enough in order to hit the weak hitbox (non-paralyze/`collision_attr_purple`).

It is literally impossible to ever get the weak hitbox (the non-paraylyze/`collision_attr_purple` one) if you let go of the special button on a `frame >= 2` and/or are too low to the ground

**IF** it is intended, the for loop can be reworked, with regards to the weak hitbox (the non-paraylyze/`collision_attr_purple` one), since:
1. Only the very **LAST** hitbox will either be strong or weak (only this hitbox would require `use_weak_hitbox`)
 a. IFF you are high enough to let `GANON_DOWN_SPECIAL_AIR_COUNTDOWN_INT` fall between 0 and 10
2. The other hitboxes don't have to check for `use_weak_hitbox` and can just use the strong hitbox values

Basically, in its current state, you will always be hitting with the strong hitbox.

**Review by FatherOfEgg — CHANGES_REQUESTED** [review]

File: src/ganon/acmd/specials/special_air_side_fall.rs
Comment (ZeDespo) [reply src/ganon/acmd/specials/special_air_side_fall.rs:1]: > Is this the intended mechanic? 
Yes it is. The idea is that when the button is held, Ganondorf is committing to the down special air. If the button is let go while in the air, Ganondorf will slow down to a stop. It is in this slow-down state where it wouldn't make sense to have the normal (strong) hitboxes. Once you let go of the speial button, 15 frames pass where the move proceeds as normal, then after that, there's a ten frame window when the weak hitboxes should come out. 

It is also wholly intentional for the strong hitbox to always come out when doing a sequence like Jump -> Down Special. You think it might be a good idea to remove the 15 frame buffer (or greatly reduce it to a number like 5) before the countdown to start the slow descent / weak hitboxes?

Also regarding point 2c, you're absolutely right, that is an oversightthat should be addressed.

Comment (FatherOfEgg) [reply src/ganon/acmd/specials/special_air_side_fall.rs:1]: > Once you let go of the special button, 15 frames pass where the move proceeds as normal, then after that, there's a ten frame window when the weak hitboxes should come out

So if everything is intended, and we do fix point 2c, it shouldn't be as difficult to hit the weak hitbox? For simplicity sake, let's say Ganon is moving at a speed of 10 units, and slowing down towards 0. Around 4 units and below, that's when the weak hitbox activates? It still feels like the weak hitbox doesn't have a purpose and wouldn't really be used.

It seems that in most situations, you're just going to either continue holding the special button to fully commit, or let go of the special button to go into the slowdown state. However, due to how the slowdown state works, you're more likely to hit with the strong hitbox anyway and never the weak. Or you could be too low (not having enough height and time to allow the countdown to fall in the range of 0 to 10 for it to change to the weak hitbox).

The weak hitbox doesn't really have a purpose, since it activates so late. You're going to hit with the strong hitbox. This move's **actual** flowchart is:
1. If you continue to hold the special button, do strong htibox, and you landed, do shockwave
2. If you let go of the special button, **STILL** have the strong hitbox, but don't do shockwave. As well as having the potential to combo the opponent when they get popped up (because the paralyze hitstun is over). And _maybe_ have a weak hitbox under a specific conditions.

There's another scenario where you are high up in the air, you hold the special button for a while, and then let go to go into the slowdown state. You can still potentially combo with flowchart point 2. Like, why wouldn't you just keep holding the special button? If you do want to let go, you can, and still have the strong hitbox be active. There's no "downside". It's always the strong hitbox. Also, you might not let go of the button high/early enough to allow the weak hitbox to activate. You might let go too "late" and still hit with the strong hitbox.

If the sequence `Jump -> Down Special`, and the like exists, is there a need for the weak hitbox? If there's no need, it can simplify some things. If we want to still incorporate the weak hitbox, we would have to rethink this move's flowchart and how the special button affects the move.

Comment (ZeDespo) [reply src/ganon/acmd/specials/special_air_side_fall.rs:1]: We could remove the weak hitbox entirely, but let's rethink it as a sour spot hitbox. If Ganondorf moves vertically at -10 units per frame, then slows down to -5 units per frame during the slowdown, that punch of his would only have half vertical velocity and therefore would not have the same amount of force behind it. The weak hitbox's sole purpose is to demonstrate that detail. 

The purpose of my implementation was to give another mixup to Ganondorf's aerial mobility. It was my intent to make the strong hitboxes to always come out when Ganondorf is low enough to the ground; however, if we had a move that is absolutely committal (like Simon's down air), it becomes predictable. So a player can either 

1) Wait until they are above stage, then commit to the down special air move with enough vertical height to trigger the "strong" version of the move. 
2) Use down special air off stage and let go of the special button at a point in time where the opponent can be crossed up. Example: if the opponent is on the ledge waiting to hit you, you can use down special air to quickly get below the ledge and use up air to drag them off the stage and spike them. 

To address point 2C and to explore this further, what if we make the rules the inverse? Instead of waiting 15 frames to slow down from the dive, we wait for 8 and the next 17 frames would have the sourspot hitbox?
